pkg/checks/checks.go

```diff
@@ -0,0 +1,186 @@
+package checks
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	appclient "k8s.io/client-go/kubernetes/typed/apps/v1"
+)
+
+// WaitForHTTPStatusOk poll until URL returns 200
+func WaitForHTTPStatusOk(ctx context.Context, transport http.RoundTripper, urltocheck string) error {
+	cli := &http.Client{
+		Transport: transport,
+		Timeout:   10 * time.Second,
+	}
+
+	req, err := http.NewRequest("GET", urltocheck, nil)
+	if err != nil {
+		return err
+	}
+
+	for {
+		resp, err := cli.Do(req)
+		if err, ok := err.(*url.Error); ok && (err.Timeout() || err.Err == io.EOF || err.Err == io.ErrUnexpectedEOF) {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		if err == io.EOF {
+			time.Sleep(time.Second)
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode == http.StatusOK {
+			return nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(time.Second):
+		}
+	}
+}
+
+func checkNamespace(namespace string) bool {
+	if namespace == "default" {
+		return true
+	}
+
+	whitelistedNamespaces := []string{
+		"openshift-",
+		"kube-",
+	}
+	for _, ns := range whitelistedNamespaces {
+		if strings.HasPrefix(namespace, ns) {
+			return true
+		}
+	}
+	return false
+}
+
+// WaitForInfraServices verify daemonsets, statefulsets
+func WaitForInfraServices(ctx context.Context, appclient *appclient.AppsV1Client) error {
+
+	// Daemonsets
+	dsList, err := appclient.DaemonSets("").List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, ds := range dsList.Items {
+		if !checkNamespace(ds.Namespace) {
+			continue
+		}
+		for {
+			ds, err := appclient.DaemonSets(ds.Namespace).Get(ds.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if ds.Status.NumberMisscheduled > 0 {
+				return fmt.Errorf("Daemonset[%v] in Namespace[%v] has missscheduled[%v]", ds.Name, ds.Namespace, ds.Status.NumberMisscheduled)
```

Comment:
I wouldn't error out here at once but retry for some time until it converges to zero.

Reply:
We can reuse the existing context timeout as a retry mechanism so you can just log here.

Reply:
Should I log or just remove this statement entirely and let the loop occur until we are satisfied with the valid logic?

Reply:
Removing is also probably fine. I would add a log that prints all status fields every time we retry but can be done in a follow-up across all loops if we want it.